src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs

```diff
@@ -59,4 +59,182 @@ private bool IsMainWindow(IntPtr handle)
+                        //
+                        // Also, EnumProcessModules is not a reliable method to get the modules for a process. 
+                        // If OS loader is touching module information, this method might fail and copy part of the data.
+                        // This is no easy solution to this problem. The only reliable way to fix this is to
+                        // suspend all the threads in target process. Of course we don't want to do this in Process class.
+                        // So we just to try avoid the race by calling the same method 50 (an arbitrary number) times.
+                        //
+                        if (!enumResult)
+                        {
+                            bool sourceProcessIsWow64 = false;
+                            bool targetProcessIsWow64 = false;
+                            SafeProcessHandle hCurProcess = SafeProcessHandle.InvalidHandle;
+                            try
+                            {
+                                hCurProcess = ProcessManager.OpenProcess(unchecked((int)Interop.Kernel32.GetCurrentProcessId()), Interop.Advapi32.ProcessOptions.PROCESS_QUERY_INFORMATION, true);
+                                bool wow64Ret;
+
+                                wow64Ret = Interop.Kernel32.IsWow64Process(hCurProcess, ref sourceProcessIsWow64);
+                                if (!wow64Ret)
+                                {
+                                    throw new Win32Exception();
+                                }
+
+                                wow64Ret = Interop.Kernel32.IsWow64Process(processHandle, ref targetProcessIsWow64);
+                                if (!wow64Ret)
+                                {
+                                    throw new Win32Exception();
+                                }
+
+                                if (sourceProcessIsWow64 && !targetProcessIsWow64)
+                                {
+                                    // Wow64 isn't going to allow this to happen, the best we can do is give a descriptive error to the user.
+                                    throw new Win32Exception(Interop.Errors.ERROR_PARTIAL_COPY, SR.EnumProcessModuleFailedDueToWow);
+                                }
+                            }
+                            finally
+                            {
+                                if (hCurProcess != SafeProcessHandle.InvalidHandle)
+                                {
+                                    hCurProcess.Dispose();
+                                }
+                            }
+
+                            // If the failure wasn't due to Wow64, try again.
+                            for (int i = 0; i < 50; i++)
+                            {
+                                enumResult = Interop.Kernel32.EnumProcessModules(processHandle, moduleHandlesArrayHandle.AddrOfPinnedObject(), moduleHandles.Length * IntPtr.Size, ref moduleCount);
```

Comment:
We really should be looping here to get the buffer to the right size. We should be getting the last error here as well (and throwing it out appropriately).

We should also be a little more explicit here- it would be clearer if the out was `bytesNeeded`.